internal/span/span.go

```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package span
+
+import (
+	"fmt"
+)
+
+type Number interface {
+	~int | ~uint64
+}
+
+// Span represents two points on a number line.
+type Span[T Number] struct {
+	Start, End T
+}
+
+// Check asserts that [Span.Start] < [Span.End].
```

Comment:
Not `<=`? In other words, all spans must have `length > 0`? That seems reasonable, but it might be worth saying rather than rewriting `if s.Start < s.End` in words.